Gears/script.cpp

```diff
@@ -1139,60 +1144,181 @@ void functionAShift() {
+                    if (currTotalPower <= (gearPredictStandardPower(currGear + 1) * 0.99f) && currSpeed > (getGearMinSpeed(currGear + 1) * economyCorrection)) {
+                        shiftTo(g_vehData.mGearCurr + 1, true);
+                        g_gearStates.FakeNeutral = false;
+                        g_gearStates.LastUpshiftTime = currGameTime;
+                    }
+                    else if (g_vehData.mRPM > 0.98f && currSpeedWorld > getGearMaxSpeed(currGear)) {
+                        shiftTo(g_vehData.mGearCurr + 1, true);
+                        g_gearStates.FakeNeutral = false;
+                        g_gearStates.LastUpshiftTime = currGameTime;
+                    }
+                }
+                else {
+                    auto currGearMaxSpeed = getGearMaxSpeed(currGear);
+                    auto prevGearMaxSpeed = getGearMaxSpeed(currGear - 1);
+                    if (currSpeed > (((currGearMaxSpeed - prevGearMaxSpeed) * economyCorrection) + (prevGearMaxSpeed * economyCorrection))) {
+                        shiftTo(g_vehData.mGearCurr + 1, true);
+                        g_gearStates.FakeNeutral = false;
+                        g_gearStates.LastUpshiftTime = currGameTime;
+                    }
+                }
+            }
+        }
+        // Shift down
+        if (currGear > 1) {
+            if (skidding) {
+                float currGearMinSpeed = getGearMinSpeed(currGear);
+                float speedGap = (currGearMinSpeed - getGearMinSpeed(currGear - 1)) / 5;
+                if (speedGap < 8) speedGap = 8;
+                if (currSpeedWorld < ((currGearMinSpeed - speedGap) * economyCorrection)) {
+                    shiftTo(currGear - 1, true);
+                    g_gearStates.FakeNeutral = false;
+                }
+            }
+            else {
+                float currGearMinSpeed = getGearMinSpeed(currGear);
+                float speedGap = (currGearMinSpeed - getGearMinSpeed(currGear - 1)) / 5;
+                if (speedGap < 8) speedGap = 8;
+                if (currSpeed < ((currGearMinSpeed - speedGap) * economyCorrection)) {
+                    shiftTo(currGear - 1, true);
+                    g_gearStates.FakeNeutral = false;
+                }
+            }
+        }
+    }
+}
+//For TCU
+float getGearMinSpeed(int gear) {
+    int currGear = g_vehData.mGearCurr;
+    float peakTorqueSpeed = (g_vehData.mDriveMaxFlatVel / g_vehData.mGearRatios[gear]) * 0.75f;
+    float prevGearMaxSpeed = getGearMaxSpeed(gear - 1); // however, the previous gear ratio might not happy with 0.75f
+    return gear > 1 ? (peakTorqueSpeed < prevGearMaxSpeed ? peakTorqueSpeed : prevGearMaxSpeed) : peakTorqueSpeed;
+}
+//For TCU
+float getGearMaxSpeed(int gear) {
+    if (gear < 1) return 0;
+    return (g_vehData.mDriveMaxFlatVel / g_vehData.mGearRatios[gear]) * 0.93f; // rev further more unlikely provides good effcient or performance due to power loss
```

Comment:
Please put long comments on a new line. (or just put this comment under "For TCU"